man/man1/fastq2bam.1
.\" Process this file with
.\" groff -man -Tascii bam-rmdup.1
.\"
.TH FASTQ2BAM 1 "OCTOBER 2014" Applications "User Manuals"
.SH NAME
fastq2bam \- convert fastq files to bam
.SH SYNOPSIS
.B fastq2bam [
.I option
.B ... ]
.SH DESCRIPTION
.B fastq2bam
converts FastQ or FastA files to BAM, honoring various common
conventions.  Paired end and singly or doubly indexed reads are
supported in most encodings encountered in the wild.

Most syntactic conventions employed as extensions to FastQ are
recognized, see below.  Multiple files can be given as input, they are
combined sensibly and finally concatenated.  Bare file arguments behave
the same as if they were given with the
.I --read-one
option.


.SH OPTIONS
.IP "-o, --output file"
Send output to
.I file
instead of standard output.  Absent
.IR -o ,
uncompressed BAM is piped to stdout.

.IP "-1, --read-one file"
Read 
.I file
and turn each contained sequence into a BAM record.  
.I file
may contain an arbitrary mix of paired and unpaired reads, first and
second mates, and so on.

.IP "-2, --read-two file"
Read
.I file
and interpret the contents as second mates of read pairs.  Each read is
paired up with one from the source previously given with
.I --read-one
(or stdin if no
.I --read-one
was encountered), and the pair is appropriately flagged as 1st mate and
2nd mate.

.IP "-I, --index-one file"
Read 
.I file
and interpret the contents as the first index sequence, which is
combined with reads from the source previously given with
.I --read-one
(or stdin if no
.I --read-one
was encountered).  

.IP "-J, --index-two file"
Read 
.I file
and interpret the contents as the second index sequence, which is
combined with reads from the source previously given with
.I --read-one
(or stdin if no
.I --read-one
was encountered).

.IP "-m, -merge-overlap[=QUAL]"
Attempt to merge or trim reads.  The optional quality is the cutoff at
which a merge/trim result is considered guaranteed correct and the
original read or read pair is discarded.

.IP "-q, --merge-qual QUAL"
Sets the quality below which merging is considered pointless.  A
merge/trim result with a lower quality is immediately discarded.

.IP "-v, --verbose"
Causes
.I fastq2bam 
to print a progress report to stderr.

.SH INPUT FILES

Input files can be FastQ or FastA, even a mix of the two,
optionally compressed using either
.IR gzip "(1) or " bgzip "(1)."
If the input is FastA, the output will not have quality scores
associated with the sequences, but nothing else changes.  Many
annotations in the header are recognized:

.IP \(bu 4 
A name suffix of \(lq/1\(rq or \(lq/2\(rq is turned into the first mate
or second mate flag, respectively (Illumina convention).

.IP \(bu 4
The name prefixes \(lqF_\(rq and \(lqR_\(rq are turned into the first
mate or second mate flag, respectively (legacy MPI EVAN convention).

.IP \(bu 4
The name prefix \(lqM_\(rq flags the sequence as unpaired and merged
(legacy MPI EVAN convention).

.IP \(bu 4 
The name prefix \(lqT_\(rq flags the sequence as unpaired and trimmed
(legacy MPI EVAN convention).

.IP \(bu 4
The name prefix of \(lqC_\(rq, either before or after any of the other
prefixes, is turned into the extra flag 
.IR XP:i:-1 ,
meaning the result of duplicate removal with unknown duplicate count
(legacy MPI EVAN convention, I'm really sorry for this one).

.IP \(bu 4
A nucleotide sequence separated from the name by an octothorpe
(\(lq#\(rq) is removed and treated as the first index.  Two such
sequences, separated by a comma, are treated as first and second index
(legacy convention, possibly MPI EVAN only).

.IP \(bu 4
If the first word of the description has at least four colon separated
subfields, the first is used to flag first/second mate if it has the
value 
.IR 1 " or " 2 ,
respectively, the second is interpreted as the "QC failed" flag if it
has the value
.IR Y ,
, and the fourth is used as the first index sequence.

If multiple files are read and combined, later files override the
appropriate values parsed from earlier files.  This makes it possible
to, for example, have one file containing the first mates with their index
sequences, another with the second mates with their index sequences, and
a third file that supplies the second index sequence only.


.SH OUTPUT FILES

Output is BAM.  Most information is encoded in the standard fields and
flags.  In addition, the first index sequence is placed into the 
.I XI
field with string type, its quality score into the 
.I YI
field with string type, encoded just like in FastQ files.
Likewise, the second index goes into
.IR XJ " and " YJ .
If a read is recognized as being a replacement for a cluster of
duplicates, this is encoded by setting
.I XP
to
.I -1 
with integer type.


.SH NOTES

Whenever multiple files are to be combined, they must run parallel, that
is, each file must contain sequences with the same read names in the
same order.

If only a
.I --read-one
argument is combined with one or two index files, the normal parsing
logic applies, so a mix of paired and unpaired reads, even with indices
is allowed.  The index sequences are overridden with those from the
separate input files.

If one 
.I --read-one
argument is combined with one
.I --read-two
argument, the pairing flags are forced.  While the normal parsing logic
still applies, a mix of paired and unpaired reads will not work as
desired and will probably lead to errors.

.SH MERGING

Merging of read pairs and trimming of single reads are considered
effectively the same.  Only few adapters occur in the wild, and we try
to trim any combination of these:
.IR "Genomic-R1" ", " "Genomic-R2" ", " "CL72" ", " "Multiplex-R2" ", " "Graft-P7" "."
(These are standard adapters on the Illumina platform on either
single-ended or paired-end flow cells; CL72 is the ancient DNA adapter
at MPI EVAN.)

Merging computes two qualities and stores them in the bam record:
\(lqYM\(rq (that's \fBM\fRerge quality) is the odds that the chosen
merge is wrong, \(lqYN\(rq (that's a\fBN\fRother merge quality) is the
odds that any possible merge is wrong, both are encoded in deciban (also
known as the Phred scale).

If \(lqYM\(rq is too low (less than 20, configurable with
.IR --merge-qual ),
nothing is merged and the original read pair is retained.  If \(lqYM\(rq
is very high (higher that 200, configurable with 
.IR --merge-overlap ), 
the original pair is discarded.  (At this quality, the program is
expected to make less than a single mistake in a lifetime of sequecing.)
Otherwise, the read is merged and the original pair is retained in
addition.

Modifications are logged in the \(lqFF\(rq field (that's \fBF\fRreaking
\fBF\fRlags):  The value is
.I 1 
for a trimmed read, 
.I 2
for a merged read, 
.I 3
for a merged read that is also trimmed,
and 
.I 4
for the original of a read that is also present in a merged or trimmed
version (all MPI EVAN convention).

.SH AUTHOR
Udo Stenzel <udo_stenzel@eva.mpg.de>

.SH "SEE ALSO"
.BR biohazard (7)



src/fastq2bam.hs
import Bio.Bam
import Bio.Bam.Evan ( removeWarts )
import Bio.Iteratee.ZLib
import Bio.Prelude
import Paths_biohazard_tools ( version )
import System.Console.GetOpt
import System.IO

import qualified Data.ByteString       as B
import qualified Data.ByteString.Char8 as S
import qualified Data.Vector.Generic   as V

data Opts = Opts { output  :: BamMeta -> Iteratee [BamRec] IO ()
                 , inputs  :: [Input]
                 , verbose :: Bool
                 , merge   :: Maybe Int
                 , lowqual :: Int }

defaultOpts :: Opts
defaultOpts = Opts { output  = protectTerm . pipeBamOutput
                   , inputs  = []
                   , verbose = False
                   , merge   = Nothing
                   , lowqual = 20 }

data Input = Input { _read1  :: FilePath         -- ^ file with first read (or other stuff)
                   ,  read2  :: Maybe FilePath   -- ^ optional file with second read
                   , index1  :: Maybe FilePath   -- ^ optional file with first index
                   , index2  :: Maybe FilePath   -- ^ optional file with second index
                   , lindex1 :: Int }            -- ^ length of first index contained in first read
  deriving Show

plainInput :: FilePath -> Input
plainInput fn = Input fn Nothing Nothing Nothing 0

getopts :: [String] -> ([Opts -> IO Opts], [String], [String])
getopts = getOpt (ReturnInOrder add_read1) options
  where
    options =
        [ Option "o" ["output"]         (ReqArg set_output "FILE") "Write output to FILE"
        , Option "1" ["read-one"]        (ReqArg add_read1 "FILE") "Parse FILE for anything"
        , Option "2" ["read-two"]        (ReqArg add_read2 "FILE") "Parse FILE for second mates"
        , Option "I" ["index-one"]        (ReqArg add_idx1 "FILE") "Parse FILE for first index"
        , Option "J" ["index-two"]        (ReqArg add_idx2 "FILE") "Parse FILE for second index"
        , Option "l" ["length-index-one"] (ReqArg set_lidx1 "NUM") "Read 1 ends on NUM index bases"
        , Option "m" ["merge-overlap"]   (OptArg set_merge "QUAL") "Attempt to merge or trim reads"
        , Option "q" ["merge-qual"]       (ReqArg set_qual "QUAL") "Minimum quality for merge is QUAL"
        , Option "v" ["verbose"]               (NoArg set_verbose) "Print progress information"
        , Option "h?" ["help","usage"]               (NoArg usage) "Print this helpful message" ]

    set_output "-" c = return $ c { output  = pipeBamOutput }
    set_output  fn c = return $ c { output  = writeBamFile fn }
    set_verbose    c = return $ c { verbose = True }

    set_merge Nothing  c =                    return $ c { merge   = Just 200 }
    set_merge (Just a) c = readIO a >>= \m -> return $ c { merge   = Just  m  }
    set_qual        a  c = readIO a >>= \q -> return $ c { lowqual =       q  }

    add_read1 fn c = return $ c { inputs = plainInput fn : inputs c }
    add_read2 fn c = return $ c { inputs = at_head (\i -> i { read2  = Just fn }) (inputs c) }
    add_idx1  fn c = return $ c { inputs = at_head (\i -> i { index1 = Just fn }) (inputs c) }
    add_idx2  fn c = return $ c { inputs = at_head (\i -> i { index2 = Just fn }) (inputs c) }

    set_lidx1  a c = readIO a >>= \n -> return $  c { inputs = at_head (\i -> i { lindex1 = n}) (inputs c) }

    at_head f [    ] = [ f $ plainInput "-" ]
    at_head f (i:is) = f i : is

    usage _ = do pn <- getProgName
                 let t = "Usage: " ++ pn ++ " [OPTION...]\n" ++
                         "Reads multiple FastA or FastQ files and converts them to BAM.  See manpage for details."
                 hPutStrLn stderr $ usageInfo t options
                 exitSuccess


main :: IO ()
main = do (opts, [], errors) <- getopts `fmap` getArgs
          unless (null errors) $ mapM_ (hPutStrLn stderr) errors >> exitFailure
          conf <- foldl (>>=) (return defaultOpts) opts
          pgm <- addPG (Just version)

          let eff_inputs = if null (inputs conf) then [ plainInput "-" ] else inputs conf
          when (verbose conf) $ mapM_ (hPrint stderr) eff_inputs

          foldr ((>=>) . enum_input) run (reverse eff_inputs) $
                joinI $ progress (verbose conf) $
                joinI $ concatMapStream (maybe concatDuals (mergeDuals $ lowqual conf) $ merge conf) $
                output conf (pgm mempty)


type UpToTwo a = (a, Maybe a)

one :: a -> UpToTwo a
one a = (a, Nothing)

two :: a -> a -> UpToTwo a
two a b = (a, Just b)

mapU2 :: (a -> b) -> UpToTwo a -> UpToTwo b
mapU2 f (a,b) = (f a, fmap f b)

concatDuals :: UpToTwo a -> [a]
concatDuals (a,Just  b) = [a,b]
concatDuals (a,Nothing) = [ a ]

mergeDuals :: Int -> Int -> UpToTwo BamRec -> [BamRec]
mergeDuals lowq highq (r1, Just r2) = mergeBam lowq highq default_fwd_adapters default_rev_adapters r1 r2
mergeDuals lowq highq (r1, Nothing) = trimBam lowq highq default_fwd_adapters r1

-- Enumerates a file.  Sequence and quality end up in b_seq and b_qual.
fromFastq :: (MonadIO m, MonadMask m) => FilePath -> Enumerator [BamRec] m a
fromFastq fp = enumAny fp $= enumInflateAny $= parseFastqCassava $= mapStream removeWarts
  where
    enumAny "-" = enumFd   defaultBufSize stdInput
    enumAny  f  = enumFile defaultBufSize   f

enum_input :: (MonadIO m, MonadMask m) => Input -> Enumerator [UpToTwo BamRec] m a
enum_input (Input r1 mr2 mi1 mi2 il1) = enum $= mapStream (addIdx il1)
  where
    enum = withIndex mi1 "XI" "YI" $ withIndex mi2 "XJ" "YJ" $
           maybe (fromFastq r1 $= mapStream one) (enumDual r1) mr2

addIdx :: Int -> UpToTwo BamRec -> UpToTwo BamRec
addIdx 0 brs = brs
addIdx l (br1, mbr2) = ( doext br1', fmap doext mbr2 )
  where
    l' = V.length (b_seq br1) - l

    br1'     = br1 { b_seq  = V.take l' (b_seq br1), b_qual = V.take l' (b_qual br1) }
    doext br = br  { b_exts = updateE "XI" (Text xi) $ updateE "YI" (Text yi) $ b_exts br }

    xi = S.pack . map showNucleotides . V.toList . V.drop l' $ b_seq  br1
    yi = B.pack . map ((+) 33 . unQ)  . V.toList . V.drop l' $ b_qual br1

-- Given an enumerator and maybe a filename, read index sequences from
-- the file and merge them with the numerator.
withIndex :: (MonadIO m, MonadMask m)
          => Maybe FilePath -> BamKey -> BamKey
          -> Enumerator [UpToTwo BamRec] m a -> Enumerator [UpToTwo BamRec] m a
withIndex Nothing      _    _ enum = enum
withIndex (Just fp) tagi tagq enum = mergeEnums enum (fromFastq fp) (convStream combine)
  where
    combine = do seqrecs <- lift headStream
                 idxrec  <- headStream
                 when (b_qname (fst seqrecs) /= b_qname idxrec) . error $
                        "read names do not match: " ++ shows (b_qname (fst seqrecs)) " & " ++ show (b_qname idxrec)

                 let idxseq  = S.pack $ map showNucleotides $ V.toList $ b_seq idxrec
                     idxqual = B.pack $ map   ((+33) . unQ) $ V.toList $ b_qual idxrec
                 return [ flip mapU2 seqrecs $
                        \r -> r { b_exts = (if B.null idxqual then id else insertE tagq (Text idxqual))
                                         $ insertE tagi (Text idxseq) $ b_exts r } ]

-- Enumerate dual files.  We read two FastQ files and match them up.  We
-- must make sure the names match, and we will flag everything as
-- 1st/2nd mate, no matter if the syntactic warts were present in the
-- files themselves.
enumDual :: (MonadIO m, MonadMask m)
         => FilePath -> FilePath -> Enumerator [UpToTwo BamRec] m a
enumDual f1 f2 = mergeEnums (fromFastq f1 $= mapStream one) (fromFastq f2) (convStream combine)
  where
    combine = do (firstMate, Nothing) <- lift headStream
                 secondMate           <- headStream

                 when (b_qname firstMate /= b_qname secondMate) . error $
                        "read names do not match: " ++ shows (b_qname firstMate) " & " ++ show (b_qname secondMate)

                 let qc = (b_flag firstMate .|. b_flag secondMate) .&. flagFailsQC
                     addx k = maybe id (updateE k) $ lookup k (b_exts firstMate) <|> lookup k (b_exts secondMate)
                     add_indexes = addx "XI" . addx "XJ" . addx "YI" . addx "YJ"

                 return [ two (firstMate  { b_flag = qc .|.  flagFirstMate .|. flagPaired .|. b_flag firstMate .&. complement flagSecondMate
                                          , b_exts = add_indexes $ b_exts firstMate })
                              (secondMate { b_flag = qc .|. flagSecondMate .|. flagPaired .|. b_flag secondMate .&. complement flagFirstMate
                                          , b_exts = add_indexes $ b_exts secondMate }) ]


progress :: MonadIO m => Bool -> Enumeratee [UpToTwo BamRec] [UpToTwo BamRec] m b
progress False = mapChunks id
progress True  = eneeCheckIfDonePass (icont . go 0 0)
  where
    go !_ !_ k (EOF         mx) = idone (liftI k) (EOF mx)
    go !l !n k (Chunk    [   ]) = liftI $ go l n k
    go !l !n k (Chunk as@(a:_)) = do
        let !n' = n + length as
            !nm = b_qname (fst a)
            !l' = l `max` S.length nm
        when (n .&. complement 0x1fff /= n' .&. complement 0x1fff) $ liftIO $ do
            hPutStr stderr $ "\27[K" ++
                replicate (l' - S.length nm) ' '
                ++ S.unpack nm ++ ", "
                ++ shows n' " records processed\n"
            hFlush stderr
        eneeCheckIfDonePass (icont . go l' n') . k $ Chunk as


